pkg/kubelet/cadvisor/util.go

```diff
@@ -40,19 +40,19 @@ func CapacityFromMachineInfo(info *cadvisorapi.MachineInfo) v1.ResourceList {
 		v1.ResourceCPU: *resource.NewMilliQuantity(
 			int64(info.NumCores*1000),
 			resource.DecimalSI),
-		v1.ResourceMemory: *resource.NewQuantity(
-			int64(info.MemoryCapacity),
-			resource.BinarySI),
 	}
 
+	totalHugePagesBytes := int64(0)
 	// if huge pages are enabled, we report them as a schedulable resource on the node
 	for _, hugepagesInfo := range info.HugePages {
 		pageSizeBytes := int64(hugepagesInfo.PageSize * 1024)
 		hugePagesBytes := pageSizeBytes * int64(hugepagesInfo.NumPages)
+		totalHugePagesBytes += hugePagesBytes
 		pageSizeQuantity := resource.NewQuantity(pageSizeBytes, resource.BinarySI)
 		c[v1helper.HugePageResourceName(*pageSizeQuantity)] = *resource.NewQuantity(hugePagesBytes, resource.BinarySI)
 	}
 
+	c[v1.ResourceMemory] = *resource.NewQuantity(int64(info.MemoryCapacity)-totalHugePagesBytes, resource.BinarySI)
```

Comment:
Will this result in huge pages being deducted from `node.allocatable` twice?

Already done here: https://github.com/kubernetes/kubernetes/blob/90bfd38b836a704c4b4a39490df868b156175e41/pkg/kubelet/nodestatus/setters.go#L398

I think it is better to keep capacity as is, and only change the allocatable on the cgroup.